client/client.go

```diff
@@ -1122,91 +1120,121 @@ func (r *NotaryRepository) bootstrapClient(checkInitialized bool) (*TUFClient, e
+	if err := r.rootFileKeyChange(cl, role, changelist.ActionCreate, pubKeys); err != nil {
 		return err
 	}
 	return r.publish(cl)
 }
 
-// Given a set of new keys to rotate to and a set of keys to drop, returns the list of current keys to use
-func (r *NotaryRepository) pubKeyListForRotation(role data.RoleName, serverManaged bool, newKeys []string) (pubKeyList data.KeyList, err error) {
-	var pubKey data.PublicKey
+// RotateKey removes all existing keys associated with the role. If no keys are
+// specified in keyList, then this creates and adds one new key or delegates
+// managing the key to the server. If key(s) are specified by keyList, then they are
+// used for signing the role.
+// These changes are staged in a changelist until publish is called.
+func (r *NotaryRepository) RotateKey(role data.RoleName, serverManagesKey bool, keyList []string) error {
+	return rotateKey(r, role, serverManagesKey, keyList, nil)
+}
 
-	// If server manages the key being rotated, request a rotation and return the new key
-	if serverManaged {
+// RotateKeyWithCert removes all existing keys associated with the role.
+// If no certificates are specified in certs then throw an error.
+// the key and certificate pairs are used for signing the role.
+// These changes are staged in a changelist until publish is called.
+func (r *NotaryRepository) RotateKeyWithCert(role data.RoleName, serverManagesKey bool, keyList []string, certs []data.PublicKey) error {
+	return rotateKey(r, role, serverManagesKey, keyList, certs)
+}
+
+// pubKeyListForRotation rotates keys with the provided key pairs. If no keys are provided then create a new key pair.
+// Returns the public key(s) of the used key pair(s).
+func (r *NotaryRepository) pubKeyListForRotation(role data.RoleName, serverManaged bool, privKeys []string, pubKeys data.KeyList) (pubKeyList data.KeyList, err error) {
+	var pubKey data.PublicKey
+	switch {
+	case serverManaged:
+		// If server manages the key being rotated, request a rotation and return the new key
 		remote := r.getRemoteStore()
 		pubKey, err = rotateRemoteKey(role, remote)
-		pubKeyList = make(data.KeyList, 0, 1)
 		pubKeyList = append(pubKeyList, pubKey)
-		if err != nil {
-			return nil, fmt.Errorf("unable to rotate remote key: %s", err)
+		return pubKeyList, err
+	case len(pubKeys) == 0 && len(privKeys) == 0:
+		if pubKey, err = r.CryptoService.Create(role, r.gun, data.ECDSAKey); err != nil {
+			return nil, fmt.Errorf("unable to generate key: %s", err)
 		}
-		return pubKeyList, nil
-	}
-
-	// If no new keys are passed in, we generate one
-	if len(newKeys) == 0 {
-		pubKeyList = make(data.KeyList, 0, 1)
-		pubKey, err = r.CryptoService.Create(role, r.gun, data.ECDSAKey)
 		pubKeyList = append(pubKeyList, pubKey)
-	}
-	if err != nil {
-		return nil, fmt.Errorf("unable to generate key: %s", err)
+	case len(privKeys) != 0:
+		if pubKeyList, err = r.pubKeyFromPrivKIDs(privKeys, pubKeys); err != nil {
```

Comment:
@endophage: I don't think it's being ignored, `pubKeyFromPrivKIDs` only returns pubkeys that match private keys?